File: bin/native/bend2/effs/tcp_poll.js
// TCP
// ===

// TCP.poll(sock, max, ms) is recv with a deadline: a recv that finds
// nothing parks on the socket and on the clock, whichever fires first;
// past the deadline it answers None{}, else Some{data} ("" is the peer's
// close, as TCP.recv answers it).
function tcp_poll(socket, max, ms, k) {
  const sys = io_sys();
  const fd = socket;
  const b = new Uint8Array(Math.max(Number(max), 1));
  const at = performance.now() + Number(ms);
  const go = () => {
    const n = Number(sys.recv(fd, sys.ptr(b), Number(max), 0));
    if (n >= 0) {
      return io_tup(socket, io_done({ $: "Some", value: io_text(b, n) }));
    }
    const code = sys.errno();
    if (code !== (sys.mac ? 35 : 11)) {
      return io_tup(socket, io_fail(code));
    }
    if (performance.now() >= at) {
      return io_tup(socket, io_done({ $: "None" }));
    }
    io_park_on(fd, false, k, go, at);
    return undefined;
  };
  return go();
}

File: bin/native/bend2/effs/tcp_poll.c
// TCP
// ===

// TCP.poll(sock, max, ms) is recv with a deadline: the park waits on the
// socket and on the clock, whichever fires first. A wake that finds data
// answers Some{data} ("" is the peer's close, as TCP.recv answers it); one
// that finds nothing parks again until the deadline, then answers None{}.
static Term tcp_poll_end(Env e, IoWork* w, Term r) {
  free(w->data);
  return io_tup(e, io_hand(w->hand), r);
}

static Term tcp_poll_more(Env e, IoWork* w) {
  int fd  = (int)w->hand;
  u64 at  = io_wait_time(w);
  w->size = io_sys_end(w, recv(fd, w->data, (size_t)w->made, 0));
  if (w->code == EAGAIN) {
    return io_tick() < at ? io_wait_on(w, fd, POLLIN, at, tcp_poll_more)
      : tcp_poll_end(e, w, io_done(e, term_pak(CID_NONE, 0)));
  }
  return tcp_poll_end(e, w, w->code ? io_fail(e, w->code, NULL) : io_done(e,
    io_box(e, CID_SOME, io_str(e, w->data, w->size))));
}

Term tcp_poll_run(Env e, Term* f, IoWork* w) {
  w->hand = (intptr_t)io_hand_v(f[0]);
  w->made = f[1] < INT32_MAX ? (intptr_t)f[1] : INT32_MAX;
  w->data = io_mem(malloc((size_t)w->made + 1));
  return io_wait_on(w, (int)w->hand, POLLIN,
    io_tick() + (u64)f[2] * 1000000ull, tcp_poll_more);
}

static void __attribute__((constructor)) tcp_poll_use(void) {
  io_eff(CID_TCP_POLL, tcp_poll_run, 0);
}
